kie-server-parent/kie-server-remote/kie-server-client/src/main/java/org/kie/server/client/jms/AsyncResponseHandler.java

```diff
@@ -0,0 +1,161 @@
+            logger.debug("Message listener for async message retrieval successfully registered on consumer {}", consumer);
+
+        } catch( JMSException jmse ) {
+            throw new KieServicesException("Unable to retrieve JMS response from queue " + responseQueue + " with selector " + selector, jmse);
+        }
+
+        ServiceResponse messageSentResponse = new ServiceResponse(ServiceResponse.ResponseType.NO_RESPONSE, "Message sent");
+        return new ServiceResponsesList(Arrays.asList(messageSentResponse));
+    }
+
+    @Override
+    public void dispose(Connection connection, Session session) {
+        // no op as the resources are closed from within message listener (via separate thread)
+    }
+
+    private class AsyncMessageListener implements MessageListener  {
+        private String selector;
+        private MessageConsumer consumer;
+        private Marshaller marshaller;
+        private KieServicesClient owner;
+
+        private Connection connection;
+        private Session session;
+
+        public AsyncMessageListener(Connection connection, Session session, String selector, MessageConsumer consumer, Marshaller marshaller, KieServicesClient owner) {
+            this.selector = selector;
+            this.consumer = consumer;
+            this.marshaller = marshaller;
+            this.owner = owner;
+            this.connection = connection;
+            this.session = session;
+        }
+
+        @Override
+        public void onMessage(Message message) {
+            try {
+                ((KieServicesClientImpl) owner).setConversationId(message.getStringProperty(JMSConstants.CONVERSATION_ID_PROPERTY_NAME));
+
+                String responseStr = ((TextMessage) message).getText();
+                logger.debug("Received response from server '{}'", responseStr);
+
+                ServiceResponsesList cmdResponse = marshaller.unmarshall(responseStr, ServiceResponsesList.class);
+                logger.debug("Unmarshalled response from async delivery {} calling callback {}", cmdResponse, callback);
+
+                callback.onResponse(selector, cmdResponse);
+                logger.debug("Callback {} successfully invoked with response {}", callback, cmdResponse);
+            } catch (Exception e) {
+                logger.error("Error while receiving message due to {}, this means response from the server won't be delivered to client", e.getMessage(), e);
+            } finally {
+                if (consumer != null) {
+                    try {
+                        consumer.close();
+                    } catch (JMSException e) {
+                        logger.warn("Error when closing JMS consumer due to {}", e.getMessage());
+                    }
+                }
+                // submit work to executor service to close resources
+                // as they cannot be closed from message listener
+                executorService.submit(new Runnable() {
```

Comment:
dispose is invoked before the message arrives as it's async. Imagine situation where you call client.startProcess that uses async handler, this means the message is sent and startProcess method completes - it does not wait for response so if we would close connection and session in dispose that would essentially kill the message listener and it would never receive the response.

does it make sense?